File: 1482. Minimum Number of Days to Make m Bouquets/Solution.py
class Solution(object):
    def minDays(self, bloomDay, m, k):
        """
        :type bloomDay: List[int]
        :type m: int
        :type k: int
        :rtype: int
        """
        if((m*k)>len(bloomDay)):
            return -1

        left = min(bloomDay)
        right = max(bloomDay)

        ans = right

        while left<=right:
            mid = left + (right-left)//2

            if(self.isPossible(bloomDay,mid,m,k)):
                ans = mid
                right = mid-1
            else:
                left = mid+1
        return ans


    def isPossible(self,bloomDay,d,m,k):
        b = 0
        count = 0
        for i in bloomDay:
            if i<=d:
                count+=1
            else:
                count = 0
            if count==k:
                b+=1
                count = 0
                if b>=m:
                    return True
        return b>=m


        

File: 1482. Minimum Number of Days to Make m Bouquets/Solution.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        if ((long) m * k > bloomDay.length) {
            return -1;
        }

        int min = Integer.MAX_VALUE;
        int max = Integer.MIN_VALUE;
        for (int day : bloomDay) {
            if (day < min) min = day;
            if (day > max) max = day;
        }

        int left = min;
        int right = max;
        int ans = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (isPossible(bloomDay, mid, m, k)) {
                ans = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return ans;
    }

    private boolean isPossible(int[] bloomDay, int d, int m, int k) {
        int p = 0;
        int count = 0;
        for (int day : bloomDay) {
            if (day <= d) {
                count++;
                if (count == k) {
                    p++;
                    count = 0;
                    if (p >= m) return true;
                }
            } else {
                count = 0;
            }
        }
        return p >= m;
    }
}